### DISCUSSION TITLE
How do I represent this scenario in keycloak?

### DISCUSSION THREAD

**Discussion by m1212e** [keycloak/keycloak#23111] [Q&A]
Hi, I can't figure out how to represent the following scenario in Keycloak:

I have domains, users, workspaces and roles in a project that needs to use Keycloak for auth.

Domains are containers for sharing roles, workspaces and users. These would be Keycloak realms.
Then Users, also pretty easy, these are the users in the realm.
Then workspaces. Workspaces represent project like things. Could be "NewWebsite" or "KeycloakImplementation". These can have users assigned and if the user is assigned, they can optionally also have a role. Workspaces are contained inside a domain.
Roles are also contained inside a domain. Whenever a user is assigned to a workspace, they can have a roles assigned too. So user A can be "ProjectManager" in the workspace "NewWebsite". But also just could be a member without a roles assigned. 

When Issuing the token this information needs to be included somehow. But I can't really figure out how to do this, Keycloak is giving me a tough time.

Thanks for all hints/suggestions on how to tackle this!

**Comment by souravs17031999** [comment]
Did you looked at protocol mappers where you can map different kinds of information dynamically about the identity in your token ?
https://www.keycloak.org/docs/latest/server_admin/#_protocol-mappers
